smithy-model/src/main/java/software/amazon/smithy/model/knowledge/TextIndex.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.model.knowledge;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.loader.Prelude;
+import software.amazon.smithy.model.node.Node;
+import software.amazon.smithy.model.node.ObjectNode;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.traits.ReferencesTrait;
+import software.amazon.smithy.model.traits.Trait;
+import software.amazon.smithy.model.validation.validators.TraitValueValidator;
+import software.amazon.smithy.utils.SmithyUnstableApi;
+
+@SmithyUnstableApi
+public final class TextIndex implements KnowledgeIndex {
+    private List<TextInstance> textInstanceList;
+
+    public TextIndex(Model model) {
+        textInstanceList = new ArrayList<>();
+        Set<String> visitedNamespaces = new HashSet<>();
+        Node validatePreludeNode = model.getMetadata().get(TraitValueValidator.VALIDATE_PRELUDE);
+        boolean validatePrelude = validatePreludeNode != null
+                ? validatePreludeNode.expectBooleanNode().getValue()
+                : false;
+
+        model.shapes().filter(shape -> validatePrelude || !Prelude.isPreludeShape(shape)).forEach(shape -> {
+            visitedNamespaces.add(shape.getId().getNamespace());
```

Comment:
`visitedNamespaces` is a set so it's necessary to avoid generating validation events multiple times for the same namespace. I can perform the de-dup of namespaces while adding unique ones to be validated into the same loop.